core/packages/contracts/src/SovereignTreasury.sol

```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.19;
+
+import "openzeppelin/access/AccessControl.sol";
+import "./IVault.sol";
+
+contract SovereignTreasury is AccessControl {
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+    bytes32 public constant SENDER_ROLE = keccak256("SENDER_ROLE");
+
+    IVault public vault;
+
+    struct Message {
+        Action action;
+        bytes payload;
+    }
+
+    enum Action {
+        // Withdraw from sovereign account and transfer to recipient.
+        // Parachain teams will occasionally send this message to retrieve collected fees.
+        Withdraw
+    }
+
+    struct WithdrawPayload {
+        address payable recipient;
+        uint256 amount;
+    }
+
+    constructor(IVault _vault) {
+        _grantRole(ADMIN_ROLE, msg.sender);
```

Comment:
ADMIN_ROLE is set to deployer account, and they can change sender role even after deployment? Should have a revokeRole there in the contract / deployment process?

Reply:
See my latest PR to support contract upgrades:

https://github.com/Snowfork/snowbridge/pull/801

I revoke the deployer account's admin privileges.